auacm/app/modules/submission_manager/judge.py

```diff
@@ -0,0 +1,225 @@
+import itertools
+import os
+import os.path
+import shlex
+import subprocess
+import threading
+import time
+
+from app import app
+
+
+ALLOWED_EXTENSIONS = ['java', 'c', 'cpp', 'py', 'go']
+
+COMPILE_COMMAND = {
+    "java": "javac {0}.java",
+    "py": None,
+    "c": "gcc {0}.c -o {0}",
+    "cpp": "g++ {0}.cpp -o {0}",
+    "go": "go build -o {0} {0}.go"
+}
+RUN_COMMAND = {
+    "java": "java -cp {0}/ {1}",
+    "py": "python {0}/{1}.py",
+    "c": "{0}/{1}",
+    "cpp": "{0}/{1}",
+    "go": "{0}/{1}"
+}
+TIMEOUT_MULTIPLIER = {
+    "java": 1.5,
+    "py": 2,
+    "c": 1,
+    "cpp": 1,
+    "go": 1
+}
+PENDING = 0
+COMPILATION_ERROR = 1
+COMPILATION_SUCCESS = 2
+RUNTIME_ERROR = 3
+TIMELIMIT_EXCEEDED = 4
+WRONG_ANSWER = 5
+CORRECT_ANSWER = 6
+
+
+def allowed_filetype(filename):
+    return ("." in filename and
+            filename.rsplit(".", 1)[1].lower() in ALLOWED_EXTENSIONS)
+
+
+def directory_for_submission(submission):
+    return os.path.join(
+        app.config["DATA_FOLDER"], "submits", str(submission.job))
+
+
+def directory_for_problem(submission):
+    return os.path.join(app.config["DATA_FOLDER"], "problems", submission.pid)
+
+
+def evaluate(submission, uploaded_file):
+    '''Attempts to compile (if necessary) then execute a given file.
+
+    :param submission: the newly created submission
+    :param uploaded_file: the uploaded file
+    :return: the status of the submission (one of the status constants above)
+    '''
+    directory = directory_for_submission(submission)
+    os.mkdir(directory)
+    uploaded_file.save(os.path.join(directory, uploaded_file.filename))
+    status = compile_submission(submission, uploaded_file)
+    if status == COMPILATION_SUCCESS:
+        status = execute_submission(submission, uploaded_file)
+    return status
+
+
+def compile_submission(submission, uploaded_file):
+    '''Compile the submission.'''
+    directory = directory_for_submission(submission)
+    filename = uploaded_file.filename
+    name, ext = filename.rsplit(".", 1)
+    # Don't compile file types that we can't compile.
+    if COMPILE_COMMAND[ext] is None:
+        return COMPILATION_SUCCESS
+    result = subprocess.call(
+        shlex.split(COMPILE_COMMAND[ext].format(os.path.join(directory, name))),
+        stderr=open(os.path.join(directory, "error.txt"), "w")
+    )
+    if result == 0:
+        return COMPILATION_SUCCESS
+    else:
+        submission.update_status('compile')
+        submission.emit_status("compile", -1)
+        return COMPILATION_ERROR
+
+
+def execute_submission(submission, uploaded_file):
```

Comment:
Vertical whitespace is nice. Since this method is kind of a monster, it would be nice if we could get some vertical whitespace separating the logical breaks.